### ISSUE TITLE
Consider `x <> y` not-equals syntax

### ISSUE THREAD

**Issue by nigeltao** [carbon-language/carbon-lang#1455]
It's currently `x != y`. https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0702.md#alternative-symbols considered `x /= y` and `x =/= y` but does not discuss `x <> y`, as in "x is less than or greater than y".

Doing so diverges from C++, but Carbon already uses `not` instead of `!` and using `<>` leads to one less meaning for the `!` symbol.

It admittedly is possibly ambiguous with "empty angle brackets", if angle brackets are or will be a Carbon thing (although I suspect they won't be to avoid ambiguity with `<<` bit-shift operators).

**Comment by SetZero** [comment]
> Doing so diverges from C++, but Carbon already uses `not` instead of `!` and using `<>` leads to one less meaning for the `!` symbol.

Just a small sidenote, but `not` as an alternative to `!` is valid C++: https://en.cppreference.com/w/cpp/keyword/not

**Comment by Turbine1991** [comment]
> > Doing so diverges from C++, but Carbon already uses `not` instead of `!` and using `<>` leads to one less meaning for the `!` symbol.
> 
> Just a small sidenote, but `not` as an alternative to `!` is valid C++: https://en.cppreference.com/w/cpp/keyword/not

I'm mindblown by this revelation! In all these years, I never knew this.

**Comment by chandlerc** [comment]
(marking this as an actual question for leads as you're somewhat asking for a decision)

My perspective is that this is covered in the linked proposal's alternatives sufficiently. Only one of the three disadvantages doesn't apply equally to `<>`:

> Disadvantages:
>
> - Unfamiliar to C++ programmers.
> - -a /= b would likely be expected to mean an a = a / b compound assignment.-
> Breaks consistency with Python, which uses not for logical negation and != for inequality comparison.

Plus a new one is added:

- Confusing when not-equal is semantically different from being either greater-than or less-than. For example, with floating point types there would be cases where two values would be not equal, but neither would be greater or less than the other.

Fundamentally, I don't think we should conflate *ordering* (relational comparisons) with *equality*. While the use of `!` is unfortunate here, I think it is small enough and unsurprising enough for users that we should keep it.

**Comment by zygoloid** [comment]
> - Breaks consistency with Python, which uses not for logical negation and != for inequality comparison.

Random aside: old versions of Python permitted `<>` as a synonym for `!=`.

> - Confusing when not-equal is semantically different from being either greater-than or less-than. 

I think for me this is a very important point. `<>` looks like "less than or greater than", and presumably wouldn't mean that for types that are not totally ordered.

Another consideration is that in C++, the `==` / `!=` operator set typically denote extensional equality, whereas `<`, `<=`, `>`, and `>=` might provide only a strict weak ordering; for example, there are C++ types for which `a == b` is a refinement of `a <= b && b <= a`. (And with C++20 comparisons, the `==` set of operators is kept distinct from the `<=>` set of operators.) Keeping the equality and relational operators a little separated in Carbon seems to improve our compatibility with C++.

I think the choice of `!=` is worth reconsidering -- `!=` is the only context in Carbon where `!` means "not" -- but I'm not yet convinced that `<>`  is a sufficient improvement to outweigh its downsides.

**Comment by chandlerc** [comment]
> I think the choice of `!=` is worth reconsidering

Should we do that now, or close this?

**Comment by emma-miler** [comment]
`!=`: is both fairly universal and implies an a checking of equality. 
`<>`: usage of the `<` and `>` characters implies some sort of ordering, and might be confusing.
`/=`: might cause confusion with the `/=` operator in other languages, in which it means "divide and set equal"
`=/=`: cumbersome to write and personally i find it hard to parse.

Personally i prefer `!=` as it is essentially standard.

**Comment by zygoloid** [comment]
We've reached consensus here: we'll stick with `!=` for now. There is a real concern that this is the only usage of `!` to mean "not" in Carbon, but we don't seem to have an alternative in front of us that has sufficient merit to outweigh the cost of picking something that will diverge from people's expectations, and in particular the downsides of `<>` seem to outweigh the benefits for both our readability goal and our C++ compatibility goal.
